flink-python/pyflink/fn_execution/boot.py

```diff
@@ -0,0 +1,352 @@
+    client = ProvisionServiceStub(channel=channel)
+    info = client.GetProvisionInfo(GetProvisionInfoRequest(), metadata=metadata).info
+    options = json_format.MessageToJson(info.pipeline_options)
+
+staged_dir = os.path.join(semi_persist_dir, "staged")
+files = []
+
+# download files
+with grpc.insecure_channel(artifact_endpoint) as channel:
+    client = ArtifactRetrievalServiceStub(channel=channel)
+    # get file list via retrieval token
+    response = client.GetManifest(GetManifestRequest(retrieval_token=info.retrieval_token),
+                                  metadata=metadata)
+    artifacts = response.manifest.artifact
+    # download files and check hash values
+    for artifact in artifacts:
+        name = artifact.name
+        files.append(name)
+        permissions = artifact.permissions
+        sha256 = artifact.sha256
+        file_path = os.path.join(staged_dir, name)
+        if os.path.exists(file_path):
+            with open(file_path, "rb") as f:
+                sha256obj = hashlib.sha256()
+                sha256obj.update(f.read())
+                hash_value = sha256obj.hexdigest()
+            if hash_value == sha256:
+                logging.info("file exist and has same sha256 hash value, skip...")
+                continue
+            else:
+                os.remove(file_path)
+        if not os.path.exists(os.path.dirname(file_path)):
+            os.makedirs(os.path.dirname(file_path), 0o755)
+        stream = client.GetArtifact(
+            GetArtifactRequest(name=name, retrieval_token=info.retrieval_token), metadata=metadata)
+        with open(file_path, "wb") as f:
+            sha256obj = hashlib.sha256()
+            for artifact_chunk in stream:
+                sha256obj.update(artifact_chunk.data)
+                f.write(artifact_chunk.data)
+            hash_value = sha256obj.hexdigest()
+        if hash_value != sha256:
+            raise Exception("sha256 not consist!")
+        os.chmod(file_path, permissions)
+
+acceptable_whl_specs = [".whl"]
+
+
+def find_beam_sdk_whl():
+    """
+    Find apache beam sdk wheel package in downloaded files. If found, it will be installed.
+
+    :return: the file name of apache beam sdk wheel package.
+    """
+    for filename in files:
+        if filename.startswith("apache_beam"):
+            for acceptable_whl_spec in acceptable_whl_specs:
+                if filename.endswith(acceptable_whl_spec):
+                    logging.info("Found Apache Beam SDK wheel: %s" % filename)
+                    return filename
+    return ""
+
+
+def python_location():
+    """
+    Returns the python executable file location.
+    If the environment variable "python" exists, using its value.
+
+    :return: The python executable file location. Default value is "python".
+    """
+    if "python" in os.environ:
+        return os.environ["python"]
+    else:
+        return "python"
+
+
+def pip_command():
```

Comment:
pip_location?

Reply:
In some case it returns a command (e.g. python -m pip) instead of a pip location, maybe "pip_command" makes more sense here?